annotations/tei_utils.py

```diff
@@ -0,0 +1,585 @@
+        for ch in element:
+            process_element(ch, html_parts, element_map, xpth)
+        html_parts.append('</ul>')
+        return {'html_parts': html_parts, 'element_map': element_map}
+
+    if tg in ('item', 'label'):
+        html_parts.append(f'<li class="tei-item" data-xpath="{xpth}">')
+        if element.text:
+            html_parts.append(_escape_html(element.text))
+        for ch in element:
+            process_element(ch, html_parts, element_map, xpth)
+            if ch.tail:
+                html_parts.append(_escape_html(ch.tail))
+        html_parts.append('</li>')
+        return {'html_parts': html_parts, 'element_map': element_map}
+
+    # ───────────────────────────────────────────────────────────────
+    # Tables
+    if tg == 'table':
+        html_parts.append(f'<table class="tei-table" data-xpath="{xpth}">')
+        for ch in element:
+            process_element(ch, html_parts, element_map, xpth)
+        html_parts.append('</table>')
+        return {'html_parts': html_parts, 'element_map': element_map}
+
+    if tg in ('row', 'rowGrp'):
+        html_parts.append('<tr>')
+        for ch in element:
+            process_element(ch, html_parts, element_map, xpth)
+        html_parts.append('</tr>')
+        return {'html_parts': html_parts, 'element_map': element_map}
+
+    if tg in ('cell', 'entry'):
+        html_parts.append('<td>')
+        if element.text:
+            html_parts.append(_escape_html(element.text))
+        for ch in element:
+            process_element(ch, html_parts, element_map, xpth)
+            if ch.tail:
+                html_parts.append(_escape_html(ch.tail))
+        html_parts.append('</td>')
+        return {'html_parts': html_parts, 'element_map': element_map}
+
+    # ───────────────────────────────────────────────────────────────
+    # Footnotes
+    if tg == 'note' and element.get('place') == 'foot':
+        # create a unique id for cross-reference
+        fn_id = f"fn-{len(element_map)}"
+        marker = f'<sup id="ref-{fn_id}" class="tei-fn-ref" data-xpath="{xpth}"></sup>'
+        html_parts.append(marker)
+
+        # collect footnote text to append later (store at element_map level)
+        note_text = _extract_full_text(element)
+        element_map.append({'type': 'footnote', 'text': note_text, 'id': fn_id})
+
+        return {'html_parts': html_parts, 'element_map': element_map}
+
+    # Handle other elements generically
+    html_parts.append(f'<span class="tei-{tg}" data-xpath="{xpth}">')
+    element_map.append({'xpath': xpth, 'element_type': tg,
+                        'start_pos': len("".join(html_parts))})
+    if element.text:
+        html_parts.append(_escape_html(element.text))
+    for ch in element:
+        process_element(ch, html_parts, element_map, xpth)
+        if ch.tail:
+            html_parts.append(_escape_html(ch.tail))
+    html_parts.append(f'</span>')
+    
+    return {'html_parts': html_parts, 'element_map': element_map}
```

Comment:
this needs to be taken apart. The most elegant solutions is probably to create a method for each tag, something like `_process_cell` or `_process_item` and then dynamically call the methods given the tag.